Jofultz wip

jofultz opened the pull request:
## Type of PR

- [ ] Documentation changes
- [X] Code changes
- [X] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR
This adds initial KeyVault functionality and tests to the code base.  I want to merge them with master so that we aren't too far out of sync and so that others may leverage the work.

## Review notes

## Issues Closed or Referenced

- References #65

bartr commented on src/main/java/com/microsoft/cse/helium/app/controllers/RobotsController.java:
```diff
@@ -0,0 +1,43 @@
+            IOUtils.copy(resourceAsStream, response.getOutputStream());
+            response.flushBuffer();
+        } catch (Exception e) {
+            System.out.println("Problem with displaying robot.txt " + e.toString());
```
should use logger instead of system.out.println so we can control verbosity

should be robots.txt

jofultz replied:
That's just a stub that I haven't done anything with since putting there.  All classes will use logger instantiated as a class member:
private static final Logger logger = LoggerFactory.getLogger([current class].class);

bartr commented on src/main/java/com/microsoft/cse/helium/app/controllers/RobotsController.java:
```diff
@@ -0,0 +1,43 @@
+/*  
+@RestController
+public class RobotsController {
+    @RequestMapping(value={"/robots*.txt"}, method = RequestMethod.GET)
```
this needs to handle robots*.txt to handle the App Service startup request which adds a random into - robots1234.txt

jofultz replied:
That will be implemented when I return to robots. This PR is not for robots work.

bartr commented on src/main/java/com/microsoft/cse/helium/app/controllers/RobotsController.java:
```diff
@@ -0,0 +1,43 @@
+        InputStream resourceAsStream = null;
+        try {
+            ClassLoader classLoader = getClass().getClassLoader();
+            resourceAsStream = classLoader.getResourceAsStream("robots.txt");
```
given how small robots.txt is, would it be faster / easier to just use a string constant?

jofultz replied:
That is a possibility.  Ideally, I'd like Spring to just return the resource and for me to not need to stream it.  I'll resolve that once I'm back on robots.

bartr commented on src/main/java/com/microsoft/cse/helium/app/controllers/SecretController.java:
```diff
@@ -0,0 +1,51 @@
+
+        try{
+   
+            String secretValue = _keyVaultService.getSecret("CosmosKey");
```
should not use a key as a sample - how about CosmosDatabase instead?

jofultz replied:
It will match C#. I'm still working on the SecretController, but I wanted to make sure we had KeyVaultService available in the code base for Sanjeev.  This PR does not resolve.

bartr replied:
Understand WIP, but please change to CosmosDatabase so that we don't accidentally expose a key.

bartr commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+    private final String _keyVaultNameRegex ="^[a-zA-Z](?!.*--)([a-zA-Z0-9-]*[a-zA-Z0-9])?$";
+    private static final Logger _logger = LoggerFactory.getLogger(KeyVaultService.class);
+
+    public KeyVaultService (@Value("${helium.keyvault.name}") String keyVaultName, @Value("${helium.environment.flag}") String environmentFlag) throws Exception {
```
KeyVaultName is the standard env var

jofultz replied:
It uses KeyVaultName, look in application.properties you'll see helium.keyvault.name=${KeyVaultName}.

bartr commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+                throw(ex);
+            }
+        } else {
+            throw new IllegalArgumentException("helium.environment.flag (value='" + _environmentFlag + "') must be 'MSI' or 'CLI'.  Check ${He_EnvironmentFlag} in your environment variables.");
```
where is this exception handled? 

One of our goals is to "fail early", so the app should end here (or soon after)

jofultz replied:
It will end, because Spring cannot instantiate the Service in order to inject it.

bartr replied:
understand, but when spring boot ends with an unhandled exception, it spews hundreds of lines of stack info which makes finding the real issue hard. As long as the Developer Experience is positive ...

bartr commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+    private Boolean checkKeyVaultName(String keyVaultName){
+        Boolean validSetting = true;
+
+        if(keyVaultName.length() < 3 || keyVaultName.length() > 24){
```
you can pass KeyVaultName or the full URL

jofultz replied:
I'll adjust before the final PR for SecretController.

sanjeevdwivedi commented on pom.xml:
```diff
@@ -64,19 +66,31 @@
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-api</artifactId>
+			<version>1.7.13</version>
```
Wondering if you can indent the file. Some of the indentation seems to be off.

jofultz replied:
It was properly indented.  I think that might have been a function of the web merge tool.

jyotsnaravikumar replied:
can we delete the exclusion?

jofultz replied:
Hopefully, it is fixed post rebase.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/HeliumRouter.java:
```diff
@@ -0,0 +1,21 @@
+package com.microsoft.cse.helium;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
```
I had deleted this file since we were going with the @RestController class. Any particular reason you are bringing this back?

jofultz replied:
No, it was deleted post my branch.  I should have showed a conflict and allowed me to exclude it.  Not sure how to remove it from the PR.  I have deleted it in the latest push.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/TestHandler.java:
```diff
@@ -0,0 +1,18 @@
+package com.microsoft.cse.helium;
```
This file was additionally deleted. Don't think you intended to bring it back.

jofultz replied:
save as previous file

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/config/WebConfig.java:
```diff
@@ -0,0 +1,11 @@
+package com.microsoft.cse.helium.app.config;
```
This file is completely commented out. Does it need to be there?

jofultz replied:
It is work in progress.  It may be removed in final form for handling robots.  However, this PR does not address robots.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+    private AzureTokenCredentials _credentials;
+    private KeyVaultClient _keyVaultClient;
+
+    private final String _keyVaultNameRegex ="^[a-zA-Z](?!.*--)([a-zA-Z0-9-]*[a-zA-Z0-9])?$";
```
A one line comment on why this regex?

jofultz replied:
There is a full description above the line where the regex is used.  I can move it to the regex string assignment if preferred.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+        // validate params and create credential object
+        if(_environmentFlag.equals("MSI")) {
+            _credentials = new MSICredentials(AzureEnvironment.AZURE);
+        } else if (_environmentFlag.equals("CLI")) {
```
indentation

jofultz replied:
I'm not sure on the indentation concern.  It is all aligned.  

I don't think it is appropriate to move the strings to a resource file as they are not for display, but rather internal class operations.  My plan was to make them constants for the resolving PR.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+            throw new IllegalArgumentException("helium.environment.flag (value='" + _environmentFlag + "') must be 'MSI' or 'CLI'.  Check ${He_EnvironmentFlag} in your environment variables.");
+        }
+        
+        if(!checkKeyVaultName(_keyVaultName)) {
```
suggest change name to something like, ifNotExistKeyVault(name) { } for better readiability

jofultz replied:
The function doesn't check only for exist.  I could change it to something like validateKeyVaultName(), but I don't believe it adds any more clarity than the current name.

ref: https://www.merriam-webster.com/dictionary/check

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+    public CertificateBundle getCertificate (String certName){
+
+        return _keyVaultClient.getCertificate(getKeyVaultUri(), certName);
+
```
inconsistent use of space after return. Here it is not needed.

jofultz replied:
Removed.

sanjeevdwivedi commented on src/test/java/com/microsoft/cse/helium/app/HeliumJavaApplicationTests.java:
```diff
@@ -0,0 +1,15 @@
+/*
```
If you are inserting a unit test in another file, feel free to remove this one. This is just a sample and doesn't have much meaning.

jofultz replied:
Removed.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+@Service
+public class KeyVaultService implements IKeyVaultService
+{
+    private String _keyVaultName = "";
```
Any particular reason you are initializing them to empty when they are going to get overwritten later? Probably better to hit a "NullPointerException" than use the empty string. Makes it a bit clearer too.

jofultz replied:
It was initially unassigned.  However, I got a linter warning about unassigned which I'm no longer getting.  The class can't be instantiated without matched params which are immediately assigned and validated.  So, it doesn't much matter whether it is null, empty, or some random string.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+            try{
+                _credentials = AzureCliCredentials.create();
+            }
+            catch (IOException ex)
```
IOException is a very broad exception and covers all the way from file read to network to practically any IO. Is it possible to shrink the scope to something smaller.

jofultz replied:
IOException is declared as thrown by AzureCliCredentials.create().  Therefore, I must catch it or declare rethrowing it.

sanjeevdwivedi commented on src/main/java/com/microsoft/cse/helium/app/services/keyvault/KeyVaultService.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    public KeyBundle getKey (String keyName){
+
```
whitespace - styling

jofultz replied:
empty line removed

jofultz commented:
I've rebased and merged conflicted files.  Additionally, I have moved controllers and config classes from previous check-ins to the correct path and fixed the package declarations.
